### PR TITLE
Collection sidesheets

### PR DESCRIPTION
**Pull request by rafpaf**
part of https://github.com/metabase/metabase/issues/48099

Adds a sidesheet to the collection page

### CODE REVIEW

**Comment by rafpaf** [comment]
* **#48082** <a href="https://app.graphite.dev/github/pr/metabase/metabase/48082?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rafpaf and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by rafpaf** [comment]
Needs some tests

**Comment by iethree** [comment]
are you planning to do description + official status in this PR or in a follow-up?

File: frontend/src/metabase/ui/components/icons/Icon/icons/verified_collection.svg
Comment (rafpaf) [inline frontend/src/metabase/ui/components/icons/Icon/icons/verified_collection.svg:1]: I exported this from Figma and changed the viewBox a bit to make the icon vertically centered: rather than `0 0 14 14` it's `0 0 14 13`

File: enterprise/frontend/src/metabase-enterprise/collections/constants.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ export const REGULAR_COLLECTION: CollectionAuthorityLevelConfig = {
 export const OFFICIAL_COLLECTION: CollectionAuthorityLevelConfig = {
   type: "official",
   name: t`Official`,
-  icon: "badge",
+  icon: "verified_collection",
```
Comment (rafpaf) [inline enterprise/frontend/src/metabase-enterprise/collections/constants.ts:17]: Let's use Vitorio's new SVG for verified collections

File: e2e/test/scenarios/collections/sidesheet.cy.spec.ts
Comment (iethree) [inline e2e/test/scenarios/collections/sidesheet.cy.spec.ts:1]: could you put this spec in an existing collection spec file to avoid paying 6-10 second performance hit for this one test?

File: enterprise/frontend/src/metabase-enterprise/collections/constants.ts (hunk 1 above)
Comment (iethree) [reply enterprise/frontend/src/metabase-enterprise/collections/constants.ts:17]: let's get a design review on this. I like it, but it has impact beyond this area of the product.

Screen Shot 2024-09-26 at 2 32 49 PM

Screen Shot 2024-09-26 at 2 33 41 PM

File: frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                collection={collection}
+                handleChangeDescription={handleChangeDescription}
+              />
+            </Stack>
+            {collection.authority_level === "official" && (
+              <Group noWrap spacing="sm" pb="xs">
+                <Icon {...iconProps} />
+                <Text lh={1}>{t`Official collection`}</Text>
+              </Group>
+            )}
```
Comment (iethree) [inline frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx:82]: it's a pain, but this is enterprise code that needs to stay out of the open source codebase.

File: frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      onKeyDown={(e: React.KeyboardEvent) => {
+        if (shouldPassKeyToTextarea(e.key)) {
+          (e.currentTarget as HTMLTextAreaElement).click();
+        }
+      }}
+      onKeyUp={(e: React.KeyboardEvent) => {
+        if (!shouldPassKeyToTextarea(e.key)) {
+          (e.currentTarget as HTMLTextAreaElement).click();
+        }
+      }}
```
Comment (iethree) [inline frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx:129]: this is nice!

1) should this be part of the editabletext component?
2) either way, this behavior should be unit tested

File: frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  const iconProps = useMemo(() => {
+    const icon = PLUGIN_COLLECTIONS.getIcon({
+      ...collection,
+      model: "collection",
+    } as ObjectWithModel);
+    if (icon.color) {
+      icon.color = color(icon.color);
+    }
+    return icon;
+  }, [collection]);
```
Comment (iethree) [inline frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx:48]: this should all probably be encapsulated in an enterprise component

File: frontend/src/metabase/components/EntityIdCard/EntityIdCard.tsx [hunk 5]
```diff
@@ -50,7 +50,14 @@ export function EntityIdCard({ entityId }: { entityId: string }) {
...
         <Text>{entityId}</Text>
-        <CopyButton value={entityId} />
+        <Box
+          component={CopyButton}
+          value={entityId}
+          c="var(--mb-color-text-light)"
+          style={{ cursor: "pointer" }}
+          pos="relative"
+          top={-2}
+        />
```
Comment (iethree) [inline frontend/src/metabase/components/EntityIdCard/EntityIdCard.tsx:60]: this copy button should have a hover style that sets the color to brand, which cannot be achieved with style props alone

File: frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx
Comment (iethree) [inline frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx:1]: This looks good! Let's add some unit tests, especially testing enterprise/oss functionality.

1. oss and EE can see description and entity ids
2. only EE shows the verified badge, even if there's an authority level set in OSS

**Review by iethree — CHANGES_REQUESTED** [review]
Looks and works great! 🫶 

Let's add some more tests, and it seems like the icon changes broke a bunch of existing tests

File: frontend/src/metabase/ui/components/icons/Icon/icons/verified_collection.svg [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+<path fill-rule="evenodd" clip-rule="evenodd" d="M8 9C8 7.61929 9.11929 6.5 10.5 6.5C11.8807 6.5 13 7.61929 13 9C13 9.61856 12.7754 10.1846 12.4032 10.6211L12.869 13.1829C12.8955 13.329 12.8334 13.477 12.7105 13.5603C12.5876 13.6437 12.4272 13.6467 12.3013 13.568L10.7311 12.5867L10.5 12.5L10.2689 12.5867L8.69875 13.568C8.57284 13.6467 8.41236 13.6437 8.28949 13.5603C8.16662 13.477 8.10449 13.329 8.13105 13.1829L8.59685 10.6212C8.22466 10.1847 8 9.61858 8 9ZM10.5 7.5C9.67157 7.5 9 8.17157 9 9C9 9.82843 9.67157 10.5 10.5 10.5C11.3284 10.5 12 9.82843 12 9C12 8.17157 11.3284 7.5 10.5 7.5Z" fill="#F9CF48"/>
```
Comment (iethree) [inline frontend/src/metabase/ui/components/icons/Icon/icons/verified_collection.svg:3]: sorry I missed this initially, but we shouldn't hard code the colors here.

There should be no `fill` on the `path` elements. the parent `svg` should have `fill="currentColor"` and we should set the color programmatically in the application code.

File: frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx (hunk 4 above)
Comment (rafpaf) [reply frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx:48]: I think PLUGIN_COLLECTIONS.getIcon is one thing in OSS and another thing in EE, so this should work in both contexts - is an enterprise component needed?

Comment (iethree) [reply frontend/src/metabase/collections/components/CollectionSidesheet/CollectionSidesheet.tsx]: at least here, we only show an icon at all in an enterprise context, right?

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionCaption.tsx [hunk 7]
```diff
@@ -62,15 +62,18 @@ export const CollectionCaption = ({
...
       {(isEditable || hasDescription) && (
         <CaptionDescription
-          key={collection.id}
-          initialValue={collection.description}
+          key={
+            // Including the description in the key prevents a stale value from
+            // being stored in the state of EditableText if the collection's
+            // description is modified in another component
+            `${collection.id}-${collection.description}`
+          }
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/CollectionHeader/CollectionCaption.tsx:70]: This fixes a bug where deleting the description from the sidesheet would cause it to reappear in the under-title description on mouseover

File: frontend/src/metabase/collections/components/CollectionInfoSidebar/CollectionInfoSidebar.tsx
Comment (rafpaf) [inline frontend/src/metabase/collections/components/CollectionInfoSidebar/CollectionInfoSidebar.tsx:1]: I've renamed this from CollectionSidesheet to CollectionInfoSidebar for parity with DashboardInfoSidebar and QuestionInfoSidebar

File: frontend/src/metabase/collections/components/CollectionInfoSidebar/CollectionInfoSidebar.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+          <Stack spacing="md">
+            <Stack spacing="xs">
+              <Title lh={1} size="sm" color="text-light">
+                {t`Description`}
+              </Title>
+              <SidesheetEditableDescription
+                description={description}
+                onChange={handleChangeDescription}
+                canWrite={canWrite}
+              />
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/CollectionInfoSidebar/CollectionInfoSidebar.tsx:60]: This component is now used in all three sidesheets

File: frontend/src/metabase/common/components/EditableDescription/EditableDescription.tsx
Comment (rafpaf) [inline frontend/src/metabase/common/components/EditableDescription/EditableDescription.tsx:1]: This is used on the collection page (underneat the title) and also in `SidesheetEditableDescription`, which is used in the sidesheets

**Review by mazameli — COMMENTED** [review]
This adheres to the designs and all, but given that we only really have description and entity ID in here, it feels a bit underwhelming as a surface area, especially since we don't store revision history or last-edited-by or insights, etc. Don't get me wrong — I like that it gives us a place to put additional things in the future, and I like that it makes collections feel a bit more like other entities. I just almost wonder if we should show collection descriptions here _instead of_ on the collection page (rather than in addition to), both to give the sidebar more of a _raison d'être_ and to clean up collection pages. Curious if @vbenedetti has thoughts on this, but he's out until next Monday. Maybe @luizarakaki has thoughts.

**Comment by luizarakaki** [comment]
> This adheres to the designs and all, but given that we only really have description and entity ID in here, it feels a bit underwhelming as a surface area, especially since we don't store revision history or last-edited-by or insights, etc. Don't get me wrong — I like that it gives us a place to put additional things in the future, and I like that it makes collections feel a bit more like other entities. I just almost wonder if we should show collection descriptions here _instead of_ on the collection page (rather than in addition to), both to give the sidebar more of a _raison d'être_ and to clean up collection pages. Curious if @vbenedetti has thoughts on this, but he's out until next Monday. Maybe @luizarakaki has thoughts.

I don't have a strong opinion about description only in the sidesheet or both. I'd keep it in both places as I don't think the description is a problem there and we are revisiting the collection page soon.

**Review by iethree — CHANGES_REQUESTED** [review]
1) let's split the unit tests out into separate files to follow the existing pattern and avoid the trap of trying to test different token features in the same file and it not working properly.
2) found an edge case: you can't edit the description of "our analytics" so we should disable the input there.
3) the failing e2e test is legitimate and related to this PR

**Review by mazameli — DISMISSED** [review]
@iethree asked me to look at the icon for official collections specifically. While I like the new icon, the sizing of the SVG needs to be fixed either in Figma and/or in code, because the folder element of the new icon looks noticeably larger than the main `folder` icon, making them look strange side-by-side.

The red ruler lines here show how their respective widths differ, and the magenta box shows how their heights are different. And next to that is a screenshot where you can see the oddity in situ:

image

File: enterprise/frontend/src/metabase-enterprise/collections/constants.ts (hunk 1 above)
Comment (rafpaf) [reply enterprise/frontend/src/metabase-enterprise/collections/constants.ts:17]: `find frontend enterprise e2e -type f | xargs sed -i "s/verified_collection/official_collection/g"`
easy peasy

**Comment by rafpaf** [comment]
@mazameli I've removed the new SVG from this PR. Let's spruce up the SVG and we can ship it in a follow-up PR.

**Comment by mazameli** [comment]
> @mazameli I've removed the new SVG from this PR. Let's spruce up the SVG and we can ship it in a follow-up PR.

sounds good

File: e2e/test/scenarios/collections/collections.cy.spec.js [hunk 9]
```diff
@@ -228,6 +229,46 @@ describe("scenarios > collection defaults", () => {
     });
   });
 
+  it("should allow description to be edited in the sidesheet", () => {
```
Comment (iethree) [inline e2e/test/scenarios/collections/collections.cy.spec.js:232]: this test is still failing

File: frontend/test/__support__/ui.tsx [hunk 10]
```diff
@@ -50,6 +54,11 @@ export interface RenderWithProvidersOptions {
   withKBar?: boolean;
   withDND?: boolean;
   withUndos?: boolean;
+  /** Token features to enable.
+   *
+   * Note: To keep tests isolated from another, don't change token features between tests in the same file. */
```
Comment (iethree) [inline frontend/test/__support__/ui.tsx:59]: ❤️

**Review by iethree — APPROVED** [review]
LGTM! 🚀
